### PR TITLE
Bind group entries

### PR DESCRIPTION
**Pull request by robtfm**
# Objective

Simplify bind group creation code. alternative to (and based on) #9476

## Solution

- Add a `BindGroupEntries` struct that can transparently be used where `&[BindGroupEntry<'b>]` is required in BindGroupDescriptors.

Allows constructing the descriptor's entries as:
```rust
render_device.create_bind_group(
    "my_bind_group",
    &my_layout,
    &BindGroupEntries::with_indexes((
        (2, &my_sampler),
        (3, my_uniform),
    )),
);
```

instead of

```rust
render_device.create_bind_group(
    "my_bind_group",
    &my_layout,
    &[
        BindGroupEntry {
            binding: 2,
            resource: BindingResource::Sampler(&my_sampler),
        },
        BindGroupEntry {
            binding: 3,
            resource: my_uniform,
        },
    ],
);
```

or

```rust
render_device.create_bind_group(
    "my_bind_group",
    &my_layout,
    &BindGroupEntries::sequential((&my_sampler, my_uniform)),
);
```

instead of

```rust
render_device.create_bind_group(
    "my_bind_group",
    &my_layout,
    &[
        BindGroupEntry {
            binding: 0,
            resource: BindingResource::Sampler(&my_sampler),
        },
        BindGroupEntry {
            binding: 1,
            resource: my_uniform,
        },
    ],
);
```

the structs has no user facing macros, is tuple-type-based so stack allocated, and has no noticeable impact on compile time.

- Also adds a `DynamicBindGroupEntries` struct with a similar api that uses a `Vec` under the hood and allows extending the entries.
- Modifies `RenderDevice::create_bind_group` to take separate arguments `label`, `layout` and `entries` instead of a `BindGroupDescriptor` struct. The struct can't be stored due to the internal references, and with only 3 members arguably does not add enough context to justify itself.
- Modify the codebase to use the new api and the `BindGroupEntries` / `DynamicBindGroupEntries` structs where appropriate (whenever the entries slice contains more than 1 member).

## Migration Guide

- Calls to `RenderDevice::create_bind_group({BindGroupDescriptor { label, layout, entries })` must be amended to `RenderDevice::create_bind_group(label, layout, entries)`.
- If `label`s have been specified as `"bind_group_name".into()`, they need to change to just `"bind_group_name"`. `Some("bind_group_name")` and `None` will still work, but `Some("bind_group_name")` can optionally be simplified to just `"bind_group_name"`.

### CODE REVIEW

File: crates/bevy_render/src/renderer/render_device.rs
Comment (IceSentry) [inline crates/bevy_render/src/renderer/render_device.rs:90]: Would be nice if this could take a `&str` directly to avoid all the `Some("")` everywhere. Is it possible to maybe have a `IntoLabel` trait implemented for `&str`?

I wouldn't block the PR on that though. Just something that could be nice too.

File: crates/bevy_pbr/src/render/mesh_bindings.rs
Comment (IceSentry) [inline crates/bevy_pbr/src/render/mesh_bindings.rs:180]: It makes sense to not change this in this PR but the pattern used here and the rest of the file will be easy to remove in a future PR.

File: examples/shader/post_processing.rs
Comment (IceSentry) [inline examples/shader/post_processing.rs:195]: I'd keep the comments for each entry here because they are intended to be educational.

**Review by IceSentry — APPROVED** [review]
Alright, I went over everything it looks really good. The diffs are so nice to look at!

LGTM

**Comment by IceSentry** [comment]
Oh, I reviewed before the `IntoLabel` but that's really nice to see too!

**Comment by IceSentry** [comment]
I think you can remove the `.as_slice()` from the PR description

File: examples/shader/post_processing.rs
Comment (robtfm) [reply examples/shader/post_processing.rs:195]: good point

File: crates/bevy_render/src/render_resource/bind_group_entries.rs
Comment (JMS55) [inline crates/bevy_render/src/render_resource/bind_group_entries.rs:89]:
```suggestion
    pub fn with_indicies(indexed_resources: impl IntoIndexedBindingArray<'b, N>) -> Self {
```

File: crates/bevy_render/src/render_resource/bind_group_entries.rs
Comment (JMS55) [inline crates/bevy_render/src/render_resource/bind_group_entries.rs:211]:
```suggestion
    pub fn new_with_indicies<const N: usize>(entries: impl IntoIndexedBindingArray<'b, N>) -> Self {
```

File: crates/bevy_render/src/render_resource/bind_group_entries.rs
Comment (JMS55) [inline crates/bevy_render/src/render_resource/bind_group_entries.rs:221]:
```suggestion
    pub fn extend_with_indicies<const N: usize>(
```

**Review by JMS55 — CHANGES_REQUESTED** [review]

File: crates/bevy_render/src/render_resource/bind_group_entries.rs
Comment (IceSentry) [reply crates/bevy_render/src/render_resource/bind_group_entries.rs:89]: wouldn't it be indices? But also, I did a quick google and it seems like indexes is also commonly used in less formal context.

Comment (JMS55) [reply crates/bevy_render/src/render_resource/bind_group_entries.rs:89]: Typo, yeah :P. Should be indices.

**Comment by robtfm** [comment]
Not sure what went wrong with the ci, I tried on my repo and it’s green. I also don’t know how to rerun it any more (rip bors)

**Review by JMS55 — APPROVED** [review]

**Comment by IceSentry** [comment]
Yeah, build works fine on my side too. Must be a CI issue, I think only maintainers are allowed to force rerun CI. Alternatively you could push a new commit

**Comment by mockersf** [comment]
I retriggered CI, all green now

File: crates/bevy_core_pipeline/src/bloom/mod.rs
Comment (mockersf) [inline crates/bevy_core_pipeline/src/bloom/mod.rs:23]: is `BindGroupEntries` not part of `*`?

Comment (robtfm) [reply crates/bevy_core_pipeline/src/bloom/mod.rs]: guess i got a little overeager with my ctrl+enter fingers

**Review by JMS55 — APPROVED** [review]

File: crates/bevy_pbr/src/render/mesh.rs
Comment (DGriffin91) [inline crates/bevy_pbr/src/render/mesh.rs:1255]: Not sure if I'm a fan of this or not. It feels like existence of the comments implies that the number should be just written in. It's really critical that these binding numbers are obvious so that 3rd party stuff knows where to bind. I especially am not sure about the `extend_sequential`.

Comment (robtfm) [reply crates/bevy_pbr/src/render/mesh.rs:1255]: I’ll change this to use the `with_indices` variants. I’ll leave the sequential api available though in case anyone wants it. 

hopefully this will change relatively soon to use automatic binding slot allocation anyway.

**Review by superdump — APPROVED** [review]
Approving but do not merge until after #9258 is merged. I don't want to disrupt that large PR.

**Comment by IceSentry** [comment]
I've been using this in a codebase that uses `UniformBuffer<T>`. I added this `impl` to make it work for me.

```rust
impl<'a, T: ShaderType + WriteInto> IntoBinding<'a> for &'a UniformBuffer<T> {
    #[inline]
    fn into_binding(self) -> BindingResource<'a> {
        BindingResource::Buffer(
            self.buffer()
                .expect("Failed to get buffer")
                .as_entire_buffer_binding(),
        )
    }
}
```

We could also add this so it can work with any buffer binding

```rust
impl<'a> IntoBinding<'a> for BufferBinding<'a> {
    #[inline]
    fn into_binding(self) -> BindingResource<'a> {
        BindingResource::Buffer(self)
    }
}
```

**Comment by robtfm** [comment]
I added your impl for `UniformBuffer` and `wgpu::BufferBinding`, thanks!

> I also have used this one for `DynamicUniformBuffer`

this one was already in the PR though

**Comment by IceSentry** [comment]
> this one was already in the PR though

Ah, right, sorry, it was in `uniform_buffer.rs`. When using it I just copied the `bind_group_entries.rs` module so I missed that one.

File: crates/bevy_gizmos/src/lib.rs
Comment (IceSentry) [inline crates/bevy_gizmos/src/lib.rs:428]: With the UniformBuffer/BufferBinding impl you should be able to use it here too now I think?

Comment (robtfm) [reply crates/bevy_gizmos/src/lib.rs:428]: i didn't use it anywhere there was only 1 entry, it doesn't really save any typing. i guess i could for consistency though?

File: crates/bevy_render/src/render_resource/uniform_buffer.rs
Comment (IceSentry) [inline crates/bevy_render/src/render_resource/uniform_buffer.rs:151]: This could also work. Up to you which one you prefer

```suggestion
        self.buffer()
            .expect("Failed to get buffer")
            .as_entire_buffer_binding()
            .into_binding(),
```

File: crates/bevy_gizmos/src/lib.rs
Comment (IceSentry) [reply crates/bevy_gizmos/src/lib.rs:428]: Right, makes sense then. I'll personally use it everywhere in my own code but I think it's fine as is.

File: crates/bevy_render/src/render_resource/uniform_buffer.rs
Comment (IceSentry) [reply crates/bevy_render/src/render_resource/uniform_buffer.rs]: Oh, sorry, the comma should have been remove 😅

File: crates/bevy_gizmos/src/lib.rs
Comment (IceSentry) [reply crates/bevy_gizmos/src/lib.rs:428]: Oof, yeah, nevermind then 😅

Comment (robtfm) [reply crates/bevy_gizmos/src/lib.rs:428]: there's actually quite a few single-entry bindings at zero over the codebase ... i'll try adding a `BindGroupEntries::single` method.

Comment (robtfm) [reply crates/bevy_gizmos/src/lib.rs]: added, i think it's nice

**Comment by IceSentry** [comment]
Here's another impl that could be useful to have

```rust
impl<'a> IntoBinding<'a> for &'a CachedTexture {
    #[inline]
    fn into_binding(self) -> BindingResource<'a> {
        BindingResource::TextureView(&self.default_view)
    }
}
```

I needed that because `ViewPrepassTextures` holds `Option<CachedTexture>` and I wanted to create a bind group with them.

**Review by superdump — APPROVED** [review]

**Pull request merged**
